File: 2517-maximum-tastiness-of-candy-basket/2517-maximum-tastiness-of-candy-basket.py
class Solution:
    def maximumTastiness(self, price: List[int], k: int) -> int:
        lo, hi = 0, 1e9
        n = len(price)
        ans = -1
        price.sort()
        while lo <= hi:
            mid = (lo + hi) // 2
            curr = price[0]
            cnt = 1
            for i in range(1, n):
                if (price[i] - curr) >= mid:
                    curr = price[i]
                    cnt += 1
            if cnt >= k:
                ans = mid
                lo = mid + 1
            else:
                hi = mid - 1
        return int(ans)

File: 2517-maximum-tastiness-of-candy-basket/2517-maximum-tastiness-of-candy-basket.cpp
using ll = long long;
class Solution {
public:
  int maximumTastiness(vector<int>& a, int k) {
    int n = a.size();
    sort(a.begin(), a.end());
    ll lo = 0, hi = 1e9;
    ll ans = -1;
    while (lo <= hi) {
      ll mid = (lo+hi)/2;
      ll cur = a[0], cnt = 1;
      for (int i=1; i<n; i++) {
        if (a[i] - cur >= mid) {
          cnt++;
          cur = a[i];
        }
      }  
      if (cnt >= k) {
        ans = mid;
        lo = mid + 1;
      }
      else hi = mid - 1;
    }
    return ans;
  }
};